contrib/graphite/blueflood.py

```diff
@@ -43,37 +78,149 @@ def __init__(self, config=None):
 
+  def complete(self, metric, query_depth):
+    metric_parts = metric.split('.')
+    if self.enable_submetrics:
+      complete_len = query_depth - 1
+    else:
+      complete_len = query_depth 
+    return (len(metric_parts) == complete_len)
 
-  def find_nodes(self, query):
-    try:
-      query_depth = len(query.pattern.split('.'))
-      #print 'DAS QUERY ' + str(query_depth) + ' ' + query.pattern
-      client = Client(self.bf_query_endpoint, self.tenant)
-      values = client.find_metrics(query.pattern)
+
+
+  def make_request(self, url, payload, headers):
+    if auth.is_active():
+      headers['X-Auth-Token'] = auth.get_token(False)
+    r = requests.get(url, params=payload, headers=headers)
+    if r.status_code == 401 and auth.is_active():
+      headers['X-Auth-Token'] = auth.get_token(True)
+      r = requests.get(url, params=payload, headers=headers)
+    return r
+
+  def find_metrics_endpoint(self, endpoint, tenant):
+    return "%s/v2.0/%s/metrics/search" % (endpoint, tenant)
+
+  def find_metrics(self, query):
+    print "BluefloodClient.find_metrics: " + str(query)
+    payload = {'query': query}
+    headers = auth.headers()
+    endpoint = self.find_metrics_endpoint(self.bf_query_endpoint, self.tenant)
+    r = self.make_request(endpoint, payload, headers)
+
+    if r.status_code == 200:
+        return [m['metric'] for m in r.json()]
+    else:
+      return []
+
+  def find_nodes_with_submetrics(self, query):
+    # By definition, when using submetrics, the names of all Leafnodes must end in a submetric alias
+    # BF doesn't know about the submetric aliases and so the aliases, (or globs corresponding to them,)
+    # must be massaged before querying BF.
+    # There are two cases above:
+
+    # 1. When you want a list of valid submetric aliases, i.e. a complete metric name followed by ".*"
+    # 2. When you have a query contains a valid submetric alias, i.e a complete metric name followed
+    #    by a valid submetric alias like "._avg"
+
+    # Every submetric leaf node is covered by one of the above two cases.  Everything else is a 
+    # branch node.
+
+    query_parts = query.pattern.split('.')
+    query_depth = len(query_parts)
+
+    # The pattern which all complete metrics will match
+    complete_pattern = '.'.join(query_parts[0:-1])
+
+    if (query_depth > 1) and (query_parts[-1] == '*'):
+      #  In this if clause we are searching for complete metric names
+      #  followed by a ".*". If so, we are requesting a list of submetric
+      #  names, so return leaf nodes for each submetric alias that is
+      #  enabled
       
-      for obj in values:
-        metric = obj['metric']
-        parts = metric.split('.')
-        metric_depth = len(parts)
-        if metric_depth > query_depth:
-          yield BranchNode('.'.join(parts[:query_depth]))
+      # First modify the pattern to get a superset that includes already complete
+      #  submetrics
+      new_pattern = complete_pattern + '*'
```

Comment:
no it isn't.  It is a superset which is what I was trying to explain in the comment.  For example, if query.pattern is "a.b.*", it won't return "a.b",  so new_pattern is "a.b*", with no second ".".  Does that make sense?  I'm trying to eliminate the second api call by combining the two.